habitat-lab/habitat/sims/habitat_simulator/sim_utilities.py

```diff
@@ -824,6 +964,58 @@ def ontop(
     return ontop_object_ids
 
 
+def on_floor(
+    sim: habitat_sim.Simulator,
+    object_a: habitat_sim.physics.ManagedRigidObject,
+    distance_threshold: float = 0.04,
+    alt_pathfinder: habitat_sim.nav.PathFinder = None,
+    island_index: int = -1,
+    ao_link_map: Dict[int, int] = None,
+    ao_aabbs: Dict[int, mn.Range3D] = None,
+) -> bool:
+    """
+    Gets whether or not the object is on the "floor" using the navmesh as an abstraction.
+    NOTE: alt_pathfinder option can be used to provide an alternative navmesh sized for objects. This would allow objects to be, for example, under tables or in corners and still be considered on the navmesh.
+
+    :param sim: The Simulator instance.
+    :param object_a: The object instance.
+    :param distance_threshold: Maximum allow-able displacement between current object position and navmesh snapped position.
+    :param alt_pathfinder:Optionally provide an alternative PathFinder specifically configured for this check. Defaults to sim.pathfinder.
+    :param island_index: Optionally limit allowed navmesh to a specific island. Default (-1) is full navmesh.
+    :param ao_link_map: A pre-computed map from link object ids to their parent ArticulatedObject's object id.
+    :param ao_aabbs: A pre-computed map from ArticulatedObject object_ids to their local bounding boxes. If not provided, recomputed as necessary.
+
+    :return: Whether or not the object is considered "on the floor" given the configuration.
+    """
+
+    assert isinstance(
+        object_a, habitat_sim.physics.ManagedRigidObject
+    ), "Object must be ManagedRigidObject, not implemented for ArticulatedObjects or links."
+
+    if alt_pathfinder is None:
+        alt_pathfinder = sim.pathfinder
+
+    assert alt_pathfinder.is_loaded
+
+    # use the object's heuristic size to estimate distance from the object center to the navmesh in order to regularize the navigability constraint for larger objects
+    obj_size, center = get_obj_size_along(
+        sim,
+        object_a.object_id,
+        mn.Vector3(0.0, -1.0, 0.0),
+        ao_link_map=ao_link_map,
+        ao_aabbs=ao_aabbs,
+    )
+
+    obj_snap = alt_pathfinder.snap_point(center, island_index=island_index)
```

Comment:
on_floor follows a very different logic than ontop. Why do we use a relaxed center point test instead of island-object contact points?

Reply:
Primarily because we don't have a unique object id for the floor vs. the wall or other architectural feature.
We could check contact points, but we'd need some logic to disambiguate those cases.
This also allows the check to be restricted to a specified island so it can semantically double as "on reachable floor".

I'm open to thoughts if you think this doesn't match what we really want it to.

Reply:
I was just wanting to understand. This makes sense, thanks.